lib/spack/spack/solver/counter.py

```diff
@@ -152,18 +152,28 @@ def possible_packages_facts(self, gen, fn):
 class FullDuplicatesCounter(MinimalDuplicatesCounter):
     def possible_packages_facts(self, gen, fn):
         build_tools = spack.repo.PATH.packages_with_tags("build-tools")
-        counter = collections.Counter(
-            list(self._link_run) + list(self._total_build) + list(self._direct_build)
+        dep_tree = spack.package_base.possible_dependencies(
+            *self.specs, virtuals=self._possible_virtuals, depflag=self.all_types
         )
-        gen.h2("Maximum number of nodes")
-        for pkg, count in sorted(counter.items(), key=lambda x: (x[1], x[0])):
-            count = min(count, 2)
-            gen.fact(fn.max_dupes(pkg, count))
+        dep_counter = collections.Counter(
+            [dep for pkg_deps in dep_tree.values() for dep in pkg_deps]
+        )
+        dupe_build_tools = set([n for n in build_tools if dep_counter[n] > 1])
+        possible_dupes = set(
+            spack.package_base.possible_dependencies(
+                *dupe_build_tools, virtuals=self._possible_virtuals, depflag=self.all_types
+            )
+        )
```

Comment:
Note: The dependent-counting is just an optimisation, I just tried to follow the logic of the previous implementation.
Having just:
```
        build_tools = spack.repo.PATH.packages_with_tags("build-tools")
        possible_dupes = set(
            spack.package_base.possible_dependencies(
                *build_tools, virtuals=self._possible_virtuals, depflag=self.all_types
            )
        )
```
instead would also work and have the same results, I believe, but would create more duplicates/unification sets (than strictly needed).